kvserver/rangefeed: remove context

wenyihu6 opened the pull request:
TODO: make sure that there is nothing spun up from stores.Rangefeed watching for streamContext since we now do not do streamCtx cancellation during per registration disconnect

Epic: none

wenyihu6 committed:
kvserver/rangefeed: remove context from kvpb.RangeFeedEventSink

This patch removes context from kvpb.RangeFeedEventSink.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/132659)
<!-- Reviewable:end -->

stevendanna committed:
kv/kvserver: remove redundant cancel func

Previously, UnbufferedSender maintained a list of cancel functions
that were canceled when processing an error. The goal of this
cancellation was to tear down running resource in the registration.

On error, something like the following would happen:

UnbufferedSender.SendBufferedError(err) would call cancel.  This
cancellation would cancel the context inherited by the registration's
runOutputLoop, causing it to exit. The registration would then call
disconnect, cancelling additional contexts, that also seem to be
intended to control the output loop, which is no longer
running. Disconnect would then send an error back to UnbufferedSender,
which would ignore it to prevent a duplicate sending of errors.

There are other error paths, where some of the components describe
above, aren't unnecessary, but, hopefully you can see that this is
ripe for some refactoring.

Here, we take a small step that we believe will also pave the way for
a clearer implementation of the BufferredSender.

Now, all error paths, start by calling
registration.Disconnect(err). The registration, cancels the context
controlling the runOutputLoop and sends an error to the stream, which
the stream then sends to the client.

There are some paths were Disconnect may be called twice still.

This simplification is made possible by one bit of indirection: The
processor now adds the registration to a map that lives on the stream
manager. This map will eventually also store any deferred cleanups
that the process requires when that stream is shutting down.

Note, we considered passing the registrations cleanup function back up
the call stack from register and may still explore that in the
future. However, it is not without challenge. Namely, the second we've
added the registration to the processor, it is possible that an error
is delivered from the processor that the stream manager would not yet
know how to handle.

We expect further naming and organizational improvements over the
coming commits.

Epic: none
Release note: None

wenyihu6 committed:
fix test crlfmt

wenyihu6 commented:
Closing - as we are splitting things out based on our new integration branch: https://github.com/cockroachdb/cockroach/pull/125872.
